tip/config.py

```diff
@@ -0,0 +1,104 @@
+import os
+import sys
+import json
+import atexit
+import inspect
+from typing import Any
+
+
+BIN_DIR = os.path.dirname(sys.argv[0])
+TIP_DIR = os.path.dirname(BIN_DIR)
+LINKS_DIR = os.path.join(TIP_DIR, "links")
+ENVIRONMENTS_DIR = os.path.join(TIP_DIR, "environments")
+CONFIG_PATH = os.path.join(TIP_DIR, "config.json")
+
+
+class Config:
+    """
+    A hierarchical configuration class that stores and retrieves settings using a dictionary-like structure.
+
+    Config allows for convenient storage and retrieval of configuration settings using a string-based hierarchical key
+    format. Keys are separated using the specified separator (default: '/'). Values can be either simple types or nested
+    dictionaries, which are automatically converted to Config instances upon retrieval. If a key does not exist, a
+    default value can be provided, which can be a factory (config will return default() in this case). In order to set
+    value that is not dumped into the file, use prefix "_".
+
+    Examples
+    --------
+    >>> config = Config({"app": {"name": "my_app", "version": "1.0"}})
+    >>> config.get("app/name")
+    'my_app'
+    >>> config.get("app/version")
+    '1.0'
+    >>> config.get("non_existent_key", "default")
+    'default'
+    >>> config.get("default_factory", lambda: 5 / 2)
+    2.5
+    >>> config['app/location'] = "path/to/server.sock"
+    >>> config.get('app/location')
+    'path/to/server.sock'
+    """
+
+    SEP = '/'
+
+    def __init__(self, config_dict: dict):
+        self._root = config_dict
+        self._tmp_root: dict[str, Any] = {}
+
+    def get(self, key, default_value=None):
+        x = self.__get_root(key)
+        try:
+            for e in key.split(self.SEP):
+                x = x[e]
+            return Config(x) if isinstance(x, dict) else x
+        except KeyError:
+            pass
+        if inspect.isfunction(default_value):
+            default_value = default_value()
+        if default_value is not None:
+            self[key] = default_value
+        return default_value
+
+    def __setitem__(self, key, value):
+        x = self.__get_root(key)
+        *dirs, file = key.split(self.SEP)
+        for e in dirs:
+            try:
+                x = x[e]
+            except KeyError:
+                x[e] = y = {}
+                x = y
+        x[file] = value
+
+    def __get_root(self, key: str) -> dict:
+        return self._tmp_root if key.startswith('_') else self._root
+
+
+def get(key, default_value=None):
+    """Get a configuration value."""
+    return _config.get(key, default_value)
+
+
+def set_value(key, value):
```

Comment:
I was very confused with naming of this function. The best way would be to have it act as `__setkey__`, but you can't do that on module level. Having name `set` is bad, because of using of the std name `set`. `__setattr__` is bad because name can include `/`. Given name isn't great because of many reasons: `set` defines the meaning of that operation pretty well; I should probably rename `get` to `get_value`; inconsistency with `Config`'s interface. If there are any ideas on how to make it better let me know.

Reply:
On the other hand we don't use set in this module, it feels pretty much finished and even if we will want to use it, we can make:

```py
make_set = set

def uses_set():
    myset = make_set()

def set(key, value): ...
```

Reply:
Ok, I've found the solution in our codebase. We can use this:

```py
def _enable_magic():
    """Enable magic methods for Config."""
    this_module = sys.modules[__name__]

    @functools.wraps(ModuleType, updated=())
    class MagicModule(ModuleType): ...

    setattr(MagicModule, "__setitem__", set_value)
    this_module.__class__ = MagicModule

_enable_magic()
```